src/identity_tree.rs

```diff
@@ -1,26 +1,247 @@
-use crate::timed_rw_lock::TimedRwLock;
+use std::{str::FromStr, sync::Arc};
+
 use semaphore::{
     merkle_tree::Hasher,
-    poseidon_tree::{PoseidonHash, PoseidonTree},
+    poseidon_tree::{PoseidonHash, PoseidonTree, Proof},
     Field,
 };
-use std::sync::Arc;
+use serde::Serialize;
+use thiserror::Error;
+use tokio::sync::RwLock;
 
 pub type Hash = <PoseidonHash as Hasher>::Hash;
 
-pub struct TreeState {
-    pub next_leaf:   usize,
-    pub merkle_tree: PoseidonTree,
+#[derive(Clone, Eq, PartialEq, Hash, Debug)]
+pub struct TreeUpdate {
+    pub leaf_index: usize,
+    pub element:    Hash,
 }
 
-pub type SharedTreeState = Arc<TimedRwLock<TreeState>>;
+impl TreeUpdate {
+    #[must_use]
+    pub const fn new(leaf_index: usize, element: Hash) -> Self {
+        Self {
+            leaf_index,
+            element,
+        }
+    }
+}
+
+struct TreeVersionData {
+    tree:      PoseidonTree,
+    diff:      Vec<TreeUpdate>,
+    next_leaf: usize,
+    next:      Option<TreeVersion>,
+}
+
+impl TreeVersionData {
+    fn empty(tree_depth: usize, initial_leaf: Field) -> Self {
+        Self {
+            tree:      PoseidonTree::new(tree_depth, initial_leaf),
+            diff:      Vec::new(),
+            next_leaf: 0,
+            next:      None,
+        }
+    }
+
+    fn next_version(&mut self) -> TreeVersion {
+        let next = TreeVersion::from(Self {
+            tree:      self.tree.clone(),
+            diff:      Vec::new(),
+            next_leaf: self.next_leaf,
+            next:      None,
+        });
+        self.next = Some(next.clone());
+        next
+    }
+
+    async fn peek_next_update(&self) -> Option<TreeUpdate> {
+        match &self.next {
+            Some(next) => {
+                let next = next.0.read().await;
+                next.diff.first().cloned()
+            }
+            None => None,
+        }
+    }
+
+    async fn apply_next_update(&mut self) {
+        if let Some(next) = self.next.clone() {
+            let mut next = next.0.write().await;
+            if let Some(update) = next.diff.first().cloned() {
+                self.update(update.leaf_index, update.element);
+                next.diff.remove(0);
+            }
+        }
+    }
+
+    fn update(&mut self, leaf_index: usize, element: Hash) {
+        self.update_without_diff(leaf_index, element);
+        self.diff.push(TreeUpdate {
+            leaf_index,
+            element,
+        });
+    }
+
+    fn update_without_diff(&mut self, leaf_index: usize, element: Hash) {
+        self.tree.set(leaf_index, element);
+        self.next_leaf = leaf_index + 1;
+    }
+}
+
+#[derive(Clone)]
```

Comment:
Derive `Debug` at least please!

Reply:
Can't, `PoseidonTree` is an upstream dep and it doesn't do `Debug`. Will change in the future, when I touch that codebase.

Reply:
Noted!